### ISSUE TITLE
Q: Get number of waiting jobs per group/all groups

### ISSUE THREAD

**Issue by slavivanov** [taskforcesh/bullmq#1200]
Hi, I ran into some issues using priority, as complexity is O(N), and we sometimes have up to 1M waiting and/or delayed jobs.
Thus, I'm considering Groups as an alternative.
Basically, the plan is to create a few groups and assign tasks to groups by their priority. 
Since the highest priority tasks are going to be much lower in number than the lower priority tasks, the round-robin approach that Groups take should work for my use case. Possibly, the highest priority tasks will be added without a group.

However, I use the number of waiting jobs to scale our workers. From the docs, it seems this is not possible with groups:
"If you only use grouped jobs in a queue, the waiting jobs list will not grow, instead it will just keep the next job to be processed if any. But you can add non-grouped jobs to the same queue, and they will get precedence from the jobs waiting in their respective groups."

My question is: Is there a way to figure out the number of waiting jobs when using groups?

Thanks!

**Comment by manast** [comment]
@slavivanov you can use  ```getGroupsJobsCount``` to get the total number of jobs that belongs to the groups  https://api.bullmq.pro/classes/Queue.html#getGroupsJobsCount
Just do not call it so frequently, even though it is quite fast it may be time-consuming if you have many groups, as it must check the length of every group (so O(N) being N the number of groups.

**Comment by slavivanov** [comment]
Thank you for the response. I wasn't clear on the question though - is
there a way to get the waiting count by group (as opposed to the
waiting+delayed count)?

**Comment by manast** [comment]
Sure, then you just use this method: https://api.bullmq.pro/classes/Queue.html#getGroupJobsCount

**Comment by slavivanov** [comment]
Perfect, thank you. I don't think I've seen these docs until now 🤦‍♂️. 
You might want to think about a more scalable way to handle jobs with priority in the future (maybe as a Pro only feature).

**Comment by manast** [comment]
Yes, priority jobs with O(logn) complexity is planned for the pro version.

**Comment by slavivanov** [comment]
Hi @manast,
I played around with Groups today. There are a few things that stood out to me:
- the compatibility class doesn't work with the Pro classes. We fixed it by rolling our own copy of the compact class, but you might want to fix this eventually.
- Waiting jobs in groups are invisible to any monitoring solution (tried with Bull Arena / Taskforce)
- Waiting jobs in groups are not retrieved with getJobCounts/getJobs which should get all jobs.

Do you have a timeline for priority with complexity O(logn)? As I'm a bit wary of breaking stuff/workflow/monitoring by moving to Groups at the moment.

Alternatively, can you let me know if Groups would work with:
- Metrics for completed, i.e. getMetrics("completed")
- Cleaning completed jobs, i.e. queue.clean(0, 10000, "completed")

**Comment by manast** [comment]
Hello @slavivanov, please see my answers below:

> the compatibility class doesn't work with the Pro classes. We fixed it by rolling our own copy of the compact class, but you might want to fix this eventually.

If you mean this one: https://api.docs.bullmq.io/classes/Queue3.html it has been deprecated and it is not supported in the Pro version and soon not be on BullMQ either. AFAIK almost nobody is using it so we won't support it anymore.

> Waiting jobs in groups are invisible to any monitoring solution (tried with Bull Arena / Taskforce)

We are about to release a small terminal UI that can be used to examine Groups, it is still Alpha but it will be released this week,  I will keep you updated.

> Waiting jobs in groups are not retrieved with getJobCounts/getJobs which should get all jobs.

Currently, in order to get the jobs belonging to groups, you have to use the specialized group getters. Is this a big inconvenience for you?

> Do you have a timeline for priority with complexity O(logn)? As I'm a bit wary of breaking stuff/workflow/monitoring by moving to Groups at the moment.

I do not have a timeline but realistically it will not be ready until this summer.-

> Alternatively, can you let me know if Groups would work with:
- Metrics for completed, i.e. getMetrics("completed")
- Cleaning completed jobs, i.e. queue.clean(0, 10000, "completed")

Yes, it should work.

**Comment by slavivanov** [comment]
Thanks for the quick response. Yes, I'd be interested in using the terminal UI to examine groups.

My main concern is not the inconvenience, but that since grouped jobs don't work with getJobCounts/getJobs(which should work for all jobs), I'm unsure what else they might not work with. In a way, grouped jobs behave in a "special" way and this can potentially cause unintended consequences.

We are still debating but will probably give it a try, and switch to prioritized jobs when they are released.

**Comment by manast** [comment]
@slavivanov we released an alpha version of the terminal UI, it should allow you to examine the jobs for now, more features coming later.
In order to install you you need to edit your ~/.npmrc file and add these lines:
```
@taskforcesh:registry=https://npm.taskforce.sh/
//npm.taskforce.sh/:_authToken=xxxxxx-xxxxxx-xxxxx-xxxx
``` 
Then install it globally with yarn for example:
```
yarn global add @taskforcesh/terminal
```

You should be able to run it with:
```
➜  ~ taskforce-terminal --help
Usage: taskforce-terminal [options]

Options:
  -V, --version                output the version number
  -q, --name [name]            Queue name (default: null)
  -p, --port [port]            redis port [6379] (default: "6379")
  --tls [tls]                  Activate secured TLS connection to Redis
  -h, --host [host]            redis host [localhost] (default: "localhost")
  -d, --database [db]          redis database [0] (default: "0")
  --passwd [passwd]            redis password
  -u, --uri [uri]              redis uri
  -s, --sentinels [host:port]  comma-separated list of sentinel host/port pairs
  -m, --master [name]          name of master node used in sentinel configuration
  --help                       display help for command
```

So for example to examine a given queue in some Redis host:
```
taskforce-terminal -q myqueuename -h my.redis.host
```
